### ISSUE TITLE
NullReferenceException in Kestrel Http1Connection

### ISSUE THREAD

**Issue by xsoheilalizadeh** [dotnet/aspnetcore#18438]
_Forwarded from [twitter][twitter-post]_

### Describe the bug

I'm working on a Proxy based on ASP.NET Core by using the [Proxykit][proxy-kit] library. This proxy is setting on a WCF client in an ASP.NET Core based microservice. While the first request from the WCF client comes to Proxy it works fine but in second request form the client, a `NullReferenceException` will occur through [`ValidateNonOriginHostHeader`][error-method] method.

#### Proxy Middlerwares

``` C#
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }
    app.UseRouting();
    app.UseEndpoints(endpoints =>
    {
        endpoints.MapHealthChecks("/health");
        endpoints.Map("/", Ok);
        endpoints.Map("/favicon.ico", Ok);
    });

    app.RunProxy<ProxyHandler>();
}
```

### Exception Stack Trace

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.ValidateNonOriginHostHeader(String hostText)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.EnsureHostHeaderExists()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TryParseRequest(ReadResult result, Boolean& endConnection)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
```

### Further technical details

[dotTraceSnapshot.zip (Line-by-line)](https://github.com/dotnet/aspnetcore/files/4080119/dotTraceSnapshot.zip)

<details>
  <summary> Console log (trace level) </summary>
  
```
dbug: Microsoft.Extensions.Hosting.Internal.Host[1]
      Hosting starting
dbug: Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer[2]
      Failed to locate the development https certificate at '(null)'.
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://localhost:4321
dbug: Microsoft.AspNetCore.Hosting.Diagnostics[0]
      Loaded hosting startup assembly Reseller.Proxy
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Code\Alibaba\Raja.Reseller\src\Reseller.Proxy
dbug: Microsoft.Extensions.Hosting.Internal.Host[2]
      Hosting started
dbug: Microsoft.AspNetCore.Server.Kestrel[39]
      Connection id "0HLSRSAM75I47" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel[1]
      Connection id "0HLSRSAM75I47" started.
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 POST http://webservices.raja.ir/online2Services.asmx text/xml; charset=utf-8 808
dbug: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[0]
      Wildcard detected, all requests with hosts will be allowed.
trce: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[2]
      All hosts are allowed.
dbug: Microsoft.AspNetCore.Routing.Matching.DfaMatcher[1000]
      No candidates found for the request path '/online2Services.asmx'
dbug: Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware[2]
      Request did not match any endpoints
dbug: Microsoft.EntityFrameworkCore.Infrastructure[10401]
      An 'IServiceProvider' was created for internal use by Entity Framework.
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.0 initialized 'ApplicationDbContext' using provider 'Microsoft.EntityFrameworkCore.SqlServer' with options: MaxPoolSize=128 MigrationsAssembly=Reseller.Proxy, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
dbug: Microsoft.EntityFrameworkCore.Query[10107]
      queryContext => ShapedQueryCompilingExpressionVisitor.SingleAsync<bool>(
          asyncEnumerable: new QueryingEnumerable<bool>(
              (RelationalQueryContext)queryContext,
              RelationalCommandCache,
              null,
              null,
              Func<QueryContext, DbDataReader, ResultContext, int[], ResultCoordinator, bool>,
              Reseller.Data.ApplicationDbContext,
              DiagnosticsLogger<Query>
          ),
          cancellationToken: queryContext.CancellationToken)
dbug: Microsoft.EntityFrameworkCore.Database.Command[20103]
      Creating DbCommand for 'ExecuteReader'.
dbug: Microsoft.EntityFrameworkCore.Database.Command[20104]
      Created DbCommand for 'ExecuteReader' (297ms).
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20000]
      Opening connection to database 'RajaReseller' on server '(localdb)\MSSQLLocalDB'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20001]
      Opened connection to database 'RajaReseller' on server '(localdb)\MSSQLLocalDB'.
dbug: Microsoft.EntityFrameworkCore.Database.Command[20100]
      Executing DbCommand [Parameters=[@__resellerId_0='?' (Size = 450)], CommandType='Text', CommandTimeout='30']
      SELECT CASE
          WHEN EXISTS (
              SELECT 1
              FROM [Resellers] AS [r]
              WHERE ([r].[ResellerId] = @__resellerId_0) AND ([r].[HasPermission] = CAST(1 AS bit))) THEN CAST(1 AS bit)
          ELSE CAST(0 AS bit)
      END
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (144ms) [Parameters=[@__resellerId_0='?' (Size = 450)], CommandType='Text', CommandTimeout='30']
      SELECT CASE
          WHEN EXISTS (
              SELECT 1
              FROM [Resellers] AS [r]
              WHERE ([r].[ResellerId] = @__resellerId_0) AND ([r].[HasPermission] = CAST(1 AS bit))) THEN CAST(1 AS bit)
          ELSE CAST(0 AS bit)
      END
dbug: Microsoft.EntityFrameworkCore.Database.Command[20300]
      A data reader was disposed.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20002]
      Closing connection to database 'RajaReseller' on server '(localdb)\MSSQLLocalDB'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20003]
      Closed connection to database 'RajaReseller' on server '(localdb)\MSSQLLocalDB'.
info: Reseller.Proxy.ProxyHandler[0]
      Forwarding to https://webservices.raja.ir/online2Services.asmx
info: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[100]
      Start processing HTTP request POST https://webservices.raja.ir/online2Services.asmx
trce: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[102]
      Request Headers:
      Cache-Control: no-cache, max-age=0
      Accept-Encoding: gzip, deflate
      Host: webservices.raja.ir
      Request-Id: |e2f15568-480f05a68467291a.4.
      SOAPAction: "http://tempuri.org/GetWagonAvaliableSeatCountWithRationArray"
      Content-Type: text/xml; charset=utf-8
      Content-Length: 808

info: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[100]
      Sending HTTP request POST https://webservices.raja.ir/online2Services.asmx
trce: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[102]
      Request Headers:
      Cache-Control: no-cache, max-age=0
      Accept-Encoding: gzip, deflate
      Host: webservices.raja.ir
      Request-Id: |e2f15568-480f05a68467291a.4.
      SOAPAction: "http://tempuri.org/GetWagonAvaliableSeatCountWithRationArray"
      Content-Type: text/xml; charset=utf-8
      Content-Length: 808

dbug: Microsoft.AspNetCore.Server.Kestrel[25]
      Connection id "0HLSRSAM75I47", Request id "0HLSRSAM75I47:00000001": started reading request body.
dbug: Microsoft.AspNetCore.Server.Kestrel[26]
      Connection id "0HLSRSAM75I47", Request id "0HLSRSAM75I47:00000001": done reading request body.
info: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[101]
      Received HTTP response after 7448.5001ms - OK
trce: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[103]
      Response Headers:
      Cache-Control: max-age=0, private
      Vary: Accept-Encoding
      Server: Microsoft-IIS/10.0
      X-AspNet-Version: 2.0.50727
      X-Powered-By: ASP.NET
      Date: Sat, 18 Jan 2020 06:50:25 GMT
      Set-Cookie: cookiesession1=2F6B9F36YCJZHO9WJAAISQEQB89W916B;Path=/;HttpOnly
      Content-Type: text/xml; charset=utf-8
      Content-Encoding: gzip
      Content-Length: 8888

info: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[101]
      End processing HTTP request after 7497.3812ms - OK
trce: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[103]
      Response Headers:
      Cache-Control: max-age=0, private
      Vary: Accept-Encoding
      Server: Microsoft-IIS/10.0
      X-AspNet-Version: 2.0.50727
      X-Powered-By: ASP.NET
      Date: Sat, 18 Jan 2020 06:50:25 GMT
      Set-Cookie: cookiesession1=2F6B9F36YCJZHO9WJAAISQEQB89W916B;Path=/;HttpOnly
      Content-Type: text/xml; charset=utf-8
      Content-Encoding: gzip
      Content-Length: 8888

dbug: Microsoft.AspNetCore.Server.Kestrel[9]
      Connection id "0HLSRSAM75I47" completed keep alive response.
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 11530.7342ms 200 text/xml; charset=utf-8
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Connection processing ended abnormally.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.ValidateNonOriginHostHeader(String hostText)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.EnsureHostHeaderExists()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TryParseRequest(ReadResult result, Boolean& endConnection)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
dbug: Microsoft.AspNetCore.Server.Kestrel[10]
      Connection id "0HLSRSAM75I47" disconnecting.
dbug: Microsoft.AspNetCore.Server.Kestrel[2]
      Connection id "0HLSRSAM75I47" stopped.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets[7]
      Connection id "0HLSRSAM75I47" sending FIN because: "The Socket transport's send loop completed gracefully."
dbug: Microsoft.AspNetCore.Server.Kestrel[39]
      Connection id "0HLSRSAM75I48" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel[1]
      Connection id "0HLSRSAM75I48" started.
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 POST http://webservices.raja.ir/online2Services.asmx text/xml; charset=utf-8 808
trce: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[2]
      All hosts are allowed.
dbug: Microsoft.AspNetCore.Routing.Matching.DfaMatcher[1000]
      No candidates found for the request path '/online2Services.asmx'
dbug: Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware[2]
      Request did not match any endpoints
info: Reseller.Proxy.ProxyHandler[0]
      Forwarding to https://webservices.raja.ir/online2Services.asmx
info: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[100]
      Start processing HTTP request POST https://webservices.raja.ir/online2Services.asmx
trce: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[102]
      Request Headers:
      Cache-Control: no-cache, max-age=0
      Accept-Encoding: gzip, deflate
      Cookie: cookiesession1=2F6B9F36YCJZHO9WJAAISQEQB89W916B
      Host: webservices.raja.ir
      Request-Id: |e2f1556a-480f05a68467291a.1.
      SOAPAction: "http://tempuri.org/GetWagonAvaliableSeatCountWithRationArray"
      Content-Type: text/xml; charset=utf-8
      Content-Length: 808

info: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[100]
      Sending HTTP request POST https://webservices.raja.ir/online2Services.asmx
trce: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[102]
      Request Headers:
      Cache-Control: no-cache, max-age=0
      Accept-Encoding: gzip, deflate
      Cookie: cookiesession1=2F6B9F36YCJZHO9WJAAISQEQB89W916B
      Host: webservices.raja.ir
      Request-Id: |e2f1556a-480f05a68467291a.1.
      SOAPAction: "http://tempuri.org/GetWagonAvaliableSeatCountWithRationArray"
      Content-Type: text/xml; charset=utf-8
      Content-Length: 808

dbug: Microsoft.AspNetCore.Server.Kestrel[25]
      Connection id "0HLSRSAM75I48", Request id "0HLSRSAM75I48:00000001": started reading request body.
dbug: Microsoft.AspNetCore.Server.Kestrel[26]
      Connection id "0HLSRSAM75I48", Request id "0HLSRSAM75I48:00000001": done reading request body.
info: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[101]
      Received HTTP response after 6366.0838ms - OK
trce: System.Net.Http.HttpClient.ProxyKitClient.ClientHandler[103]
      Response Headers:
      Cache-Control: max-age=0, private
      Vary: Accept-Encoding
      Server: Microsoft-IIS/10.0
      X-AspNet-Version: 2.0.50727
      X-Powered-By: ASP.NET
      Date: Sat, 18 Jan 2020 06:50:52 GMT
      Content-Type: text/xml; charset=utf-8
      Content-Encoding: gzip
      Content-Length: 8854

info: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[101]
      End processing HTTP request after 6368.043ms - OK
trce: System.Net.Http.HttpClient.ProxyKitClient.LogicalHandler[103]
      Response Headers:
      Cache-Control: max-age=0, private
      Vary: Accept-Encoding
      Server: Microsoft-IIS/10.0
      X-AspNet-Version: 2.0.50727
      X-Powered-By: ASP.NET
      Date: Sat, 18 Jan 2020 06:50:52 GMT
      Content-Type: text/xml; charset=utf-8
      Content-Encoding: gzip
      Content-Length: 8854

dbug: Microsoft.AspNetCore.Server.Kestrel[9]
      Connection id "0HLSRSAM75I48" completed keep alive response.
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 6383.0401ms 200 text/xml; charset=utf-8
```
</details>

### dotnet --info

```
ASP.NET Core 3.1

JetBrains Rider 2019.3 / dotnet run

.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

.NET Core SDKs installed:
  2.1.607 [C:\Program Files\dotnet\sdk]
  3.0.101 [C:\Program Files\dotnet\sdk]
  3.1.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```
According to the console log, I guess the exception occurs when kestrel keeps the first connection alive while the proxy finished the response, if I try to send the requests with only one WFC client instance this issue occurs while if recreate it will be fine.

```c#
using var client = GetClient();
var firstResult = client.GetSomething(); // Fetch result successfully
var secondResult =  client.GetSomething(); // Exception occurs in proxy by kestrel
```

```c#
using var firstClient = GetClient();
var firstResult = firstClient.GetSomething(); // Fetch result successfully
using var secondClient = GetClient();
var secondResult =  secondClient.GetSomething(); // Fetch result successfully
```

[twitter-post]:https://twitter.com/xsoheilalizadeh/status/1217432588358422529
[proxy-kit]:https://github.com/proxykit/ProxyKit
[error-method]:https://github.com/dotnet/aspnetcore/blob/deceebc062af5a51e53cf6a8cb539da118d46484/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L559

\cc @anurse @jkotalik

**Comment by anurse** [comment]
@xsoheilalizadeh Are you able to reproduce this under a debugger at all? I ask because if you can do so and download symbols for Kestrel ([docs on how to do so in VS](https://docs.microsoft.com/en-us/visualstudio/debugger/how-to-use-the-modules-window?view=vs-2019#load-symbols), they should all be on the Microsoft Symbol Servers) then we can get the exact line number of the null-ref, which would help a lot.

@halter73 @Tratcher @jkotalik Looking at [the code](https://github.com/dotnet/aspnetcore/blob/deceebc062af5a51e53cf6a8cb539da118d46484/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L559) it seems like `_absoluteRequestTarget` is the main (possibly only?) thing that could be `null`.

**Comment by jkotalik** [comment]
Yeah that's what I noticed as well. My guess is that somehow, the Http1Connection has already been reset.
Can you print off what the HttpRequestMessage sent from your client looks like as well? Maybe the client you have isn't sending valid HTTP?

**Comment by halter73** [comment]
Just knowing a NullReferenceException was possible in the ValidateNonOriginHostHeader method helped me track down the issue. I created a PR to fix it at #18515. See the PR for more details.

For now, you can work around this issue by setting [KestrelServerOptions.DisableStringReuse](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.core.kestrelserveroptions.disablestringreuse?view=aspnetcore-3.1) to true.

**Comment by xsoheilalizadeh** [comment]
Thanks for following up. @halter73 The fix PR address refers to another issue.

**Comment by Tratcher** [comment]
That should have said https://github.com/dotnet/aspnetcore/pull/18529.

**Comment by anurse** [comment]
This was fixed in https://github.com/dotnet/aspnetcore/pull/18547 which will be released in 3.1.3.
